behave_test_automation/pah_behave/features/steps/login_view.py

```diff
@@ -55,3 +55,15 @@ def click_login(context) -> None:
 def click_login(context) -> None:
     driver_page = context.driver.current_url
     assert driver_page == 'http://localhost:8080/drive'
+
+
+@given('User is logged into pah-fm website')
+def logged_in_user(context) -> None:
+    # This is a step which executes several steps in one
+    context.execute_steps("""
+        Given User opens pah-fm website
```

Comment:
This is a pure behave feature: https://behave.readthedocs.io/en/latest/api.html#docid-api-calling-steps-from-other-steps. I'm not sure why I shouldn't use it, when it's there :)
Also, I will probably make more moves like this in the feature. Since I can't tell exactly when, and to which cases, to reach a consensus, I can apply the suggested login simulation.
Nevertheless, I would like to feel free to call steps from other ones in the future. 
Sounds like a deal?

Reply:
I don't like this approach because it obscures the behave steps. The unwritten rule is that single statement is responsible for single action. User clicks, user inputs etc that's all good but then out of nowhere you have User is logged in which underneath executes multiple behave steps.

I get that you want to have it in Given so that you can rerun it for each scenario but I think that it would be better to:
1. handle this in single python function which does as described, preserving the one step rule which you already do in `User opens pah-fm website`.
2. write the behave steps explicitly in Given block
```
Feature: Logout

    Background: User is logged into pah-fm website
      Given user opens up pah-fm website
      And User inputs "hello@codeforpoznan.pl" username
      And User inputs "cfp123" password
      And User is logged in successfully
      And User clicks login button
```

Reply:
Ad 1 - If I will implement the function, nothing will change Gherkin-wise -> I will still have `User is logged in` step, which, as I can see in ad 2. is an issue.
So, ad 2. - with this approach the feature file will stop being expressive. It's easier to see that `User is logged in` then to look through all the steps leading to this particular state. I don't want to focus on the login process in the logout feature file, since it's exactly the same as in login feature file. Imagine any other new feature. Each requires from the user to be logged in. This way I will have massive Backgrounds for no reason. 
Also, since we're mentioning rules, I'll allow myself to recall the DRY one - I think that putting several steps into one, which will be used across all features is a quite nice usage of it

Reply:
> Ad 1 - If I will implement the function, nothing will change Gherkin-wise -> I will still have `User is logged in` step, which, as I can see in ad 2. is an issue.

Yes, but in the end we'll have Gherkin code (`Given`) which calls Python code (`@given`), which then goes back to Gherkin and executes its own public "API"[1] (`User inputs`) which then goes back to Python (`@step`) to execute appropriate functions implementing said steps.

That to me looks odd and might bite us later when more steps will be added.

[1] by public API I mean the behave steps - that's the interface exposed from Python implementation that's available to the real user, a list of options one can choose to interact with the app.


> So, ad 2. - with this approach the feature file will stop being expressive. It's easier to see that `User is logged in` then to look through all the steps leading to this particular state. I don't want to focus on the login process in the logout feature file, since it's exactly the same as in login feature file. Imagine any other new feature. Each requires from the user to be logged in. This way I will have massive Backgrounds for no reason.
> Also, since we're mentioning rules, I'll allow myself to recall the DRY one - I think that putting several steps into one, which will be used across all features is a quite nice usage of it

My reasoning here is that feature files are inherently very repetitive so few more lines describing the Given process won't matter in overall view of the feature file but okay, the point with not implementing logging in steps in logout feature is a rather bad idea, so I can also ditch that :dagger:.


To sum up - I'm approving this because it's working and apparently this trick is supported by behave but I'm still not too sure about it ;)